edge-modules/api-proxy-module/src/monitors/config_monitor.rs

```diff
@@ -0,0 +1,406 @@
+        max_back_off,
+        keep_alive,
+    ) {
+        Ok(client) => client,
+        Err(err) => return Err(anyhow::anyhow!("Could not create client: {:?}", err)),
+    };
+
+    Ok(client)
+}
+
+pub fn start(
+    mut client: Client,
+    notify_received_config: Arc<Notify>,
+) -> Result<(JoinHandle<Result<()>>, ShutdownHandle), Error> {
+    let shutdown_signal = Arc::new(Notify::new());
+    let shutdown_handle = ShutdownHandle(shutdown_signal.clone());
+
+    use futures_util::StreamExt;
+
+    //Set default value for some environment variables here
+    set_default_env_vars();
+
+    //Allow on level of indirection, when one env var references another env var.
+    dereference_env_variable();
+
+    //Parse default config and notify to reboot nginx if it has already started
+    //If the config is incorrect, return error because otherwise nginx doesn't have any config.
+
+    match parse_config() {
+        //Notify watchdog config is there
+        Ok(()) => notify_received_config.notify(),
+        Err(err) => {
+            return Err(anyhow::anyhow!(
+                "Error while parsing default config: {:?}",
+                err
+            ))
+        }
+    };
+
+    let monitor_loop: JoinHandle<Result<()>> = tokio::spawn(async move {
+        loop {
+            let wait_shutdown = shutdown_signal.notified();
+            futures::pin_mut!(wait_shutdown);
+
+            let message = match futures::future::select(wait_shutdown, client.next()).await {
+                Either::Left(_) => {
+                    log::warn!("Shutting down config monitor!");
+                    return Ok(());
+                }
+                Either::Right((Some(Ok(message)), _)) => message,
+                Either::Right((Some(Err(err)), _)) => {
+                    log::error!("Error receiving a message! {:?}", err);
+                    continue;
+                }
+                Either::Right((None, _)) => {
+                    log::warn!("Shutting down config monitor!");
+                    return Ok(());
+                }
+            };
+
+            if let azure_iot_mqtt::module::Message::TwinPatch(twin) = message {
+                if let Err(err) = save_raw_config(&twin) {
+                    log::error!("received message {:?}", err);
+                } else {
+                    log::info!("received config {:?}", twin);
+                    //Here we don't need to panic if config is wrong. There is already a good config running.
+                    match parse_config() {
+                        //Notify watchdog config is there
+                        Ok(()) => notify_received_config.notify(),
+                        Err(error) => {
+                            log::error!("Error while parsing default config: {:?}", error)
+                        }
+                    };
+                }
+            };
+        }
+    });
+
+    Ok((monitor_loop, shutdown_handle))
+}
+
+fn set_default_env_vars() {
```

Comment:
do you think it is a good idea to re-write a bunch of ENV?

Reply:
We discussed it over the phone